cpp/src/parquet/arrow/schema.cc

```diff
@@ -984,21 +990,51 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
   bool modified = false;
 
   auto& origin_type = origin_field.type();
+  const auto& inferred_type = inferred->field->type();
 
   if (origin_type->id() == ::arrow::Type::EXTENSION) {
     const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*origin_type);
-    auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
+    if (inferred_type->id() != ::arrow::Type::EXTENSION &&
+        ex_type.extension_name() == std::string("arrow.json")) {
+      // Schema mismatch.
+      //
+      // Arrow extensions are DISABLED in Parquet.
+      // origin_type is ::arrow::extension::json()
+      // inferred_type is ::arrow::binary()
+      //
+      // Origin type is restored as Arrow should be considered the source of truth.
+      DCHECK_EQ(inferred_type->id(), ::arrow::Type::STRING);
```

Comment:
This is coming from an arbitrary Parquet file, so we should not use DCHECK. Instead let's make it part of the condition above:
```c++
    if (inferred_type->id() == ::arrow::Type::STRING &&
        ex_type.extension_name() == std::string("arrow.json")) {
      // Schema mismatch. [...]
```